bugbug/repository.py

```diff
@@ -19,6 +21,9 @@
 
 BUG_PATTERN = re.compile('[\t ]*[Bb][Uu][Gg][\t ]*([0-9]+)')
 
+COMPONENTS = json.loads(
+    subprocess.check_output('./mach file-info bugzilla-component **/* --format json', shell=True))
```

Comment:
Is `shell=True` making the Travis build to fail? Or should I add `python` before the subprocess command?

It says:
```
subprocess.CalledProcessError: Command './mach file-info bugzilla-component **/* --format json' returned non-zero exit status 127.
The command "python run.py --train --goal bug" exited with 1.
```

Reply:
Yes, that should work. Hopefully the `components` dict will not be too big (if it is, the memory usage is going to skyrocket as it is going to be copied in all processes we spawn with ProcessPoolExecutor).

Reply:
> Moreover, this command is not generating info for all files, you should use the other one we talked about

This one?
`./mach file-info bugzilla-automation OUTDIR`

But it is producing empty json files.

Reply:
There's likely a problem if it is producing empty files, as those files should have information (otherwise the commend would be totally useless :P).
Maybe there's something wrong in your configuration. Have you tried running `./mach bootstrap` from the command line?

Reply:
> Maybe there's something wrong in your configuration. Have you tried running `./mach bootstrap` from the command line?

Same result even after running `./mach bootstrap`

Reply:
Yes, I'm running `./mach file-info bugzilla-automation OUTDIR` from the mozilla-central directory.
And it still shows `Examined 0 files`.

Reply:
But the other sub commands like `bugzilla-component`, `dep-tests`, etc. are working fine. They are showing outputs. But `bugzilla-automation` doesn't seem to parse files.

Reply:
This is very strange, as it works for me. I see `Examined 269726 files` and the components.json file is ~30 MB.

I'm not sure what could be causing this. Maybe try adding some logging to https://searchfox.org/mozilla-central/source/python/mozbuild/mozbuild/frontend/mach_commands.py#179 and try to figure out what's wrong.

Reply:
Maybe, if you are on Windows, try using WSL or a VM.

Reply:
The problem is here:
https://searchfox.org/mozilla-central/source/python/mozbuild/mozbuild/frontend/mach_commands.py#323

The path argument in `reader.finder.find(search)` has an (extra) preceeding colon which is making the directory slightly invalid which results in no files being read, making it to return an empty iterator.

Changing the argument somewhat solves the problem but now the code here:
https://searchfox.org/mozilla-central/source/python/mozbuild/mozbuild/frontend/reader.py#1242 raises an exception:
`Exception: Path outside topsrcdir: /app.mozbuild`


But I still tried to run the code by commenting out that exception part and it works fine.
`Examined 191 files`

Reply:
191 files seem too low to be true

Reply:
I think that's due to some of the files getting skipped because of that exception.

Any idea why this exception is being raised?
`Exception: Path outside topsrcdir: /app.mozbuild`

Reply:
Are you on Windows or Linux?

Reply:
Windows

Reply:
Try my earlier suggestion (using WSL or a VM with Linux).

Reply:
I tried WSL and finally, it worked.

`Examined 269233 files` and the `components.json` file is 30.4 MB.

Thanks!

Reply:
Great! Let me know when the patch is ready then :)
(You can test it by adding a limit to the hg log call, otherwise it will take too long to run)

Reply:
> We should avoid `shell=True`, it's not really needed here.

Removing `shell=True` gives an error in my terminal window. So should I keep it here or not?

Reply:
If the component of a file is missing, just ignore it and don't add it to the components_touched list.

Reply:
> Removing `shell=True` gives an error in my terminal window. So should I keep it here or not?

You should remove it, and change the `subprocess.check_output` call accordingly.

Reply:
The first change that you definitely have to do is that the command should not be a string, but a list (['mach', 'XXX', 'YYY', 'ZZZ' ...]).

Reply:
I've made the changes. Waiting for review!

Reply:
Do we also need to `cd` to `repo_dir` before calling `mach` in the subprocess?

Reply:
And now that we have `hg_log` function in `repository.py` instead to `hg.log`, where should I add the `limit=1024`?

Here?
`args = hglib.util.cmdbuilder(b'log', template=template)`

Or there is some other way around? This also needs to be changed in the `README` file.

Reply:
> Do we also need to `cd` to `repo_dir` before calling `mach` in the subprocess?

No need to `cd` to `repo_dir`, but you will need to specifiy the full path to mach (os.path.join(repo_dir, 'mach')), and you'll need specify the working directory for the command (`repo_dir`) using the subprocess parameter to select it (it's probably something like cwd).

Reply:
Regarding the limit, you can add it in cmdbuilder, yes.

Reply:
I've made the changes and the script runs successfully now. 

After running the script, each commit in `commits.json` has the required `components` feature:
```
{"author": "Sebastian Hengst <archaeopteryx@coole-files.de>", "desc": "Bug 1514594: Follow-up: Remove unused imports merged from autoland. a=eslint-fix", "date": "2019-01-29 16:02:57", "bug_id": 1514594, "added": 1, "deleted": 5, "files_modified_num": 3, "types": ["JavaScript"], "components": ["Core", "General", "DOM: Security", "WebExtensions"]}
```